### ISSUE TITLE
[Feature]: Create a "store" that Playwright Tests can use to share data between tests

### ISSUE THREAD

**Issue by ITenthusiasm** [microsoft/playwright#32354]
### 🚀 Feature Request

This is a re-opening of #28420. I believe it was mistakenly closed because the resolver concluded that the OP wanted something that could create dynamic data within a _single_ test. However, the real feature request was the ability to share dynamic data across _multiple_ Playwright tests (through _local variables_ -- not databases or the like).

### The Problem
For different reasons, 2 or more tests may be required to share data with each other. For example, perhaps Test 2 relies on data that Test 1 created. Currently, it doesn't seem to be possible to handle this use case in Playwright. (At least, this is the case when running tests in parallel. I haven't tried `serial` tests or tests that only use 1 worker. However, I'd rather keep using multiple workers and parallel tests if possible.)

### Example

As an example, this feature could enable developers to test the following:
1) User sign up succeeds with `email1`
2) User sign up fails with `email1` because the account already exists
3) User sign _in_ succeeds with `email1` because the account already exists

Similarly, this feature could be used to test resource creation, retrieval, modification, and deletion [in `serial` mode]. All of these could be tested using a _single_ resource, instead of having to create a new resource for all 4 of these tests.

### Motivation

Benefits of having this feature:
1) **Efficiency**: Time and resources are wasted when a test has to go through the process of creating new resources through database interactions -- especially if this has to be done a lot. Being able to _reuse_ data that has already been created would be more performant than always creating new data.
2) **Clarity/Maintainability**: Fixtures are a useful feature, but they're also a somewhat advanced topic. (I saw Kent get tripped up on one of the GitHub issues here.) Typically, when writing tests that interact with a server (say, in Jest or Vitest), developers don't have to write data to the filesystem or leverage a more-advanced feature when they just want to reuse data between tests. All they really need to do is create a variable that's accessible across the related tests' scopes. If developers gain the power to share data between tests _easily_, then Playwright test code becomes easier to write and read, making Playwright better.

**Comment by Skn0tt** [comment]
Hey Isaiah! Thanks for opening the feature request. Have you looked at test [steps](https://playwright.dev/docs/next/api/class-test#test-step) to separate long tests into multiple sections? That'd allow you to put all your test scenarios into the same test, but still have it somewhat readable. Here's a snippet that shows how it'd look, and how you can share data between the tests:

```ts
import { test } from "@playwright/test"

test("user creation flow", async ({ page }) => {
   const userID = await test.step("signup succeeds", async () => {
       // do your signup
       return "whatever-user-id-you-obtained"
   })
   await test.step("signup fails if email address is already used", async () => {
       // do another signup, assert that if failed
   })
   await test.step("signin succeeds", async () => {
       // do your signin, assert that it succeeds
       // you can use userID in here
   })
})
```

Because these are steps within one test, they will always be executed serially - no matter what `workers` settings you have set in Playwright, for example.

Would this existing feature help solve your usecase? If not, why?

**Comment by ITenthusiasm** [comment]
Hey Simon! That looks incredible! Obviously, that would be quite useful for CRUD interactions performed via the UI. As for a single user authentication flow, I think that would be quite helpful. However, it doesn't quite cover all of my use cases, I don't think.

There are some additional [test cases](https://github.com/ITenthusiasm/remix-supertokens/blob/main/TEST_CASES.md#authenticated-users) I'd like to check, such as what an authenticated user can do or how session refreshing behaves. For these, I'd like to be able to:
1) **Leverage an already-existing account.** I've already expressed my reasoning for this in the OP.
2) **Create distinct tests**. The user creation flow isn't related to the flow of performing a session refresh while a user is submitting data to the server. Consequently, it wouldn't make sense to group the two scenarios under the same test name.

Does that make sense? Any thoughts on that?

Also, a separate question: What does the test report look like when using the `step` approach? Would the developer see 4 Pass/Fails (1 for the parent test, and 3 for the steps), or would the developer only see 1 Pass/Fail (for the parent test only)?

**Comment by Skn0tt** [comment]
Happy you like `test.step`. I think worker-scoped fixtures are the perfect solution to your needs: https://playwright.dev/docs/test-fixtures#worker-scoped-fixtures
You mention they feel like an advanced topic. That makes me think we should write better documentation about it, instead of creating another feature to solve your needs. I'm curious, what materials did you refer to when trying to understand fixtures, and what made you think of them as being advanced? Also, what made you think they're not a good solution to your needs?

> What does the test report look like when using the step approach?

It's one test that happens to have four steps, so they'll see 1 passed or failed.

**Comment by ITenthusiasm** [comment]
> I think worker-scoped fixtures are the perfect solution to your needs

Okay cool! I can try them out and get back to you on whether or not everything works in my tests. Seems hopeful!

> I'm curious, what materials did you refer to when trying to understand fixtures, and what made you think of them as being advanced? Also, what made you think they're not a good solution to your needs?

I want to give a thoughtful answer, but the day's gotten a little chaotic. If it's okay, I'll respond to this question later today or sometime tomorrow -- Lord willing.

**Comment by LinkPovilas** [comment]
Sharing state between tests is generally considered a bad practice and an anti-pattern. There are better ways to handle such scenarios depending on the use case:

- [Playwright Global Setup and Teardown](https://playwright.dev/docs/test-global-setup-teardown)
- [Playwright Setup and Worker Fixtures](https://playwright.dev/docs/auth)
- [Playwright Test Fixtures](https://playwright.dev/docs/test-fixtures)

**Comment by ITenthusiasm** [comment]
@Skn0tt

Sorry for the delay. It _is_ technically still Thursday, but much later in the day than I wanted it to be.

> I'm curious, what materials did you refer to when trying to understand fixtures,

When I was originally introduced to fixtures in the first quarter of this year, I used Playwright's [official documentation](https://playwright.dev/docs/test-fixtures) to familiarize myself with fixtures. However, I haven't used fixtures since the first quarter, and I only used test-scoped fixtures at the time. (I did read about worker-scoped fixtures. But there were some other fixture concepts that I ended up skipping because the new concept was slightly difficult for me to understand at the time.)

> and what made you think of them as being advanced?

It's a powerful tool that solves an important testing problem -- specifically a problem that not all developers are guaranteed to encounter the first time that they start using Playwright significantly. (I ended up needing to reach for something like fixtures the 3rd or 4th time I used Playwright, and I didn't even know they existed until that point.) Additionally, they're a tool that solves the problem in a magical way, and they require more study upfront to be used effectively. Usually, tools that check these boxes (whether in testing libraries or other kinds of libraries) are "more advanced" -- at least in my mind.

Fixtures were the first Playwright concept that took me a while to (start to) understand. They're great. But the fact that they're "executed" by destructuring them in tests makes them a bit magical, and the fact that an object always has to be destructured when creating a fixture -- [even if nothing in that object is used](https://playwright.dev/docs/test-fixtures#fixture-timeout) -- is a little unintuitive. (That last point was a footgun for me. And it was also a footgun for someone else in another GitHub issue, though I don't remember which issue it was.) Moreover, if you're using multiple fixtures, you really have to understand the "order of operations" to make sure everything behaves as expected. It's not something as easy as, "Look at API, toy with `method` for 1min, then start using effectively in your tests 'forever'."

> Also, what made you think they're not a good solution to your needs?

Ignorance. I was working off of my previous memory/understanding of fixtures -- which had apparently failed me. I thought that worker-scoped fixtures wouldn't be able to accomplish what I wanted. But so far it seems like they're _working_ ( :wink: ). (I still have to confirm that the same username and password are being used for the tests that I'm writing. Up until now, I've been assuming that's the case.)

---

Does that make sense? I wanted to give a full response, but it was a bit rushed due to time constraints. Feel free to challenge any of that.

**Comment by Skn0tt** [comment]
That makes total sense, thank you for the elaborate answer. I understand that the destructuring feels magic, it's one of the behaviours I was also surprised by when I first learned about it.

They're powerful, and that makes them complex to understand. It's inherent complexity though, so we won't be able to remove with better API design. But we can do better learning materials! I'll bring this up with the team.

---

I'll go ahead and close the issue since it seems we found the solution to be worker-scoped fixtures. Good thing you opened the issue though, this discussion will be super helpful for future folks googling for this!

**Comment by ITenthusiasm** [comment]
Got it. Hopefully something can be added to the docs to help other people like me. Thanks for talking things through with me on this one, @Skn0tt!

**Comment by ITenthusiasm** [comment]
@Skn0tt Sorry for the 2nd ping. I've been using worker-scoped fixutres, but my application has been running into unexpected race conditions that cause the application to error out. This is because the same Worker Fixture is running multiple times within the same test file. From what you described earlier, I was expecting that using a Worker-scoped Fixture would force all of the tests using said fixture to belong to the same worker. (Thus, the fixture would only run once, and no race conditions would occur.) But that doesn't seem to be the case.

When logging the `workerInfo.workerIndex` inside my Worker Fixture, I saw `108` for one test and `109` for another. (Both tests were using the same worker-scoped fixture.)

- Repo: [Remix SuperTokens](https://github.com/ITenthusiasm/remix-supertokens)
- [Worker Fixture](https://github.com/ITenthusiasm/remix-supertokens/blob/main/e2e/auth.test.ts#L31-L53)
- To Test: `npm run build && npm start` in one terminal. Then `npx playwrigh test --ui` in another.

You don't necessarily have to test this locally. But if you do, you can use the ENV Vars mentioned in the `README.md`.

**Comment by Skn0tt** [comment]
Interesting! Looking at https://playwright.dev/docs/test-parallel#introduction, it says that "Tests in a single file are run [...]in the same worker process.", except when you increased parallelism manually. Could you take a look at that doc and check if it helps you? If not, we might be looking at a bug - please open another issue for that then.

**Comment by ITenthusiasm** [comment]
I see. Looking at that documentation, it looks like Playwright does not intend to share any data between workers. That's helpful to know in relationship to the OP.

Separately, it seems like the default configuration that Playwright provides uses `fullyParallel: true`, and it doesn't apply any limit to the `workers` (unless you're in a CI environment). Would that be the problem? If so, what would the recommended solution be? I'd still like to have multiple workers and run things in parallel in other test files when I create them.

It's sounding like I should use `test.describe.configure({ mode: "serial" })` for this specific test file (though the docs discourage its use)? I could also try to split out the tests that are using the Worker Fixture into their own `describe` block that uses `mode: "serial"` so that the remaining tests in the same file could be run in parallel.

**Comment by ITenthusiasm** [comment]
@Skn0tt Didn't get a response so just pinging for awareness. (If you don't prefer pinging, or if I should reach out through a different medium like Discord, let me know.)

**Comment by Skn0tt** [comment]
Sorry that I didn't get back earlier. Pinging is great!

I'm assuming you refer to the [docs on Serial mode](https://playwright.dev/docs/test-parallel#serial-mode), which is the perfect docs for your use case. The reason we discourage serial mode is that interdependent tests can be hard to work with. If one test depends on another, then how would you run only a single of them?
Dependencies between tests tend to be a sign that those tests should be a single file instead.

But Serial mode works perfectly fine, so feel free to use it if it solves your use case.
